src/openloco/windows/construction/ConstructionTab.cpp

```diff
@@ -0,0 +1,1276 @@
+                i = 0;
+                auto height = 0;
+
+                while (roadPiece[i].index != 0xFF)
+                {
+                    if (height > roadPiece[i].z)
+                        height = roadPiece[i].z;
+                    i++;
+                }
+
+                roadHeight -= height;
+                roadHeight -= 16;
+                ebx = 2;
+
+                if (input::has_key_modifier(input::key_modifier::shift))
+                {
+                    ebx = 0x80000000;
+                    roadHeight -= 16;
+                }
+            }
+            else
+            {
+                ebx = 1;
+                roadHeight = _word_1135FFE;
+            }
+
+            while (true)
+            {
+                _constructionHover = 0;
+                _byte_113607E = 0;
+                _x = mapPos.x;
+                _y = mapPos.y;
+                _word_1135FB8 = roadHeight;
+                _byte_522096 = 0;
+                _byte_1136066 = 0;
+
+                common::activateSelectedConstructionWidgets();
+                auto window = WindowManager::find(WindowType::construction);
+
+                if (window == nullptr)
+                    return;
+
+                _byte_508F09 = _byte_508F09 | (1 << 0);
+
+                on_mouse_up(window, widx::construct);
+
+                _byte_508F09 = _byte_508F09 & ~(1 << 0);
+
+                if (_dword_1135F42 == 0x80000000)
+                {
+                    if (gGameCommandErrorText != string_ids::error_can_only_build_above_ground)
+                    {
+                        ebx--;
+                        if (ebx != 0)
+                        {
+                            roadHeight -= 16;
+                            if (roadHeight >= 0)
+                            {
+                                if (ebx < 0)
+                                {
+                                    continue;
+                                }
+                                else
+                                {
+                                    roadHeight += 32;
+                                    continue;
+                                }
+                            }
+                        }
+                    }
+                }
+                else
+                {
+                    _byte_113607E = 1;
+                    WindowManager::close(WindowType::error, 0);
+                    return;
+                }
+
+                on_mouse_up(window, widx::rotate_90);
+
+                audio::play_sound(audio::sound_id::error, int32_t(input::getMouseLocation().x));
+
+                return;
+            }
+        }
+        else
+        {
+            map_invalidate_map_selection_tiles();
+            common::sub_49FEC7();
+
+            auto track = common::getTrackPieceId(_lastSelectedTrackPiece, _lastSelectedTrackGradient, _constructionRotation);
+
+            if (!track)
```

Comment:
You should be able to common up a lot of this code with the road code above. I'm not going to comment on issues that are present in both.